comms/dht/src/origin_mac.rs

```diff
@@ -0,0 +1,183 @@
+//  Copyright 2022, The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::convert::TryFrom;
+
+use digest::{Digest, FixedOutput};
+use rand::rngs::OsRng;
+use tari_comms::types::{Challenge, CommsPublicKey, CommsSecretKey, Signature};
+use tari_crypto::keys::PublicKey;
+use tari_utilities::ByteArray;
+
+#[derive(Debug, Clone)]
+pub struct OriginMac {
+    signer_public_key: CommsPublicKey,
+    signature: Signature,
+}
+
+fn construct_origin_mac_hash(
+    signer_public_key: &CommsPublicKey,
+    public_nonce: &CommsPublicKey,
+    message: &[u8],
+) -> [u8; 32] {
+    // e = H_mac(P||R||m)
+    Challenge::with_params(&[], &[], b"TARIDHTORIGINMAC")
+        .chain(signer_public_key.as_bytes())
+        .chain(public_nonce.as_bytes())
+        .chain(message)
+        .finalize_fixed()
+        .into()
+}
+
+impl OriginMac {
+    /// Create a new signed [OriginMac](self::OriginMac) for the given message.
+    pub fn new_signed(signer_secret_key: CommsSecretKey, message: &[u8]) -> Self {
+        let (nonce_s, nonce_pk) = CommsPublicKey::random_keypair(&mut OsRng);
+        let signer_public_key = CommsPublicKey::from_secret_key(&signer_secret_key);
+        let challenge = construct_origin_mac_hash(&signer_public_key, &nonce_pk, message);
+        let signature = Signature::sign(signer_secret_key, nonce_s, &challenge)
+            .expect("challenge is [u8;32] but SchnorrSignature::sign failed");
```

Comment:
I suppose this is basically infallible.

Reply:
It's 100% infallible with a 32-byte challenge and Ristretto - we should probably make an infallible version of sign that takes a [u8;32]